cpp/src/io/json/json_gpu.cu

```diff
@@ -41,56 +43,53 @@ using string_pair = std::pair<const char *, size_t>;
+ * @param[in] begin Pointer to the first character in the row
+ * @param[in] end pointer to the first character after the row
+ */
+__device__ std::pair<char const *, char const *> limit_range_to_brackets(char const *begin,
+                                                                         char const *end)
 {
-  while (start < stop && data[start] != '[' && data[start] != '{') { start++; }
-  start++;
+  while (begin < end && *begin != '[' && *begin != '{') { begin++; }
+  begin++;
 
-  while (start < stop && data[stop - 1] != ']' && data[stop - 1] != '}') { stop--; }
-  stop--;
+  while (begin < end && *(end - 1) != ']' && *(end - 1) != '}') { end--; }
+  end--;
+  return {begin, end};
 }
 
 /**
- * @brief CUDA kernel that finds the end position of the next field name,
- * including the colon that separates the name from the field value.
- *
- * Returns the position after the colon that precedes the value token.
- *
- * @param[in] data Pointer to the device buffer containing the data to process
- * @param[in] opts Parsing options (e.g. delimiter and quotation character)
- * @param[in] start Offset of the first character in the range
- * @param[in] stop Offset of the first character after the range
- *
- * @return long Position of the first character after the field name.
- **/
-__device__ long seek_field_name_end(const char *data,
-                                    const ParseOptions opts,
-                                    long start,
-                                    long stop)
+ * @brief Find the first JSON object key in the range.
+ *
+ * Assumes that begin is not in the middle of a field.
+ *
+ * @param[in] begin Pointer to the first character in the parsing range
+ * @param[in] end pointer to the first character after the parsing range
+ * @param[in] quotechar The character used to denote quotes
+ *
+ * @return Begin and end iterators of the key name; (`end`, `end`) if a key is not found
+ */
+__device__ std::pair<char const *, char const *> get_next_key(char const *begin,
+                                                              char const *end,
+                                                              char quotechar)
 {
-  bool quotation = false;
-  for (auto pos = start; pos < stop; ++pos) {
-    // Ignore escaped quotes
-    if (data[pos] == opts.quotechar && data[pos - 1] != '\\') {
-      quotation = !quotation;
-    } else if (!quotation && data[pos] == ':') {
-      return pos + 1;
-    }
-  }
-  return stop;
+  // Key starts after the first quote
+  auto const key_begin = thrust::find(thrust::seq, begin, end, quotechar) + 1;
+  if (key_begin > end) return {end, end};
+
+  // Key ends after the next unescaped quote
+  auto prev_ch       = ' ';
+  auto const key_end = thrust::find_if(thrust::seq, key_begin, end, [&] __device__(auto ch) {
+    auto const is_unescaped_quote = (ch == quotechar && prev_ch != '\\');
+    prev_ch                       = ch;
+    return is_unescaped_quote;
+  });
```

Comment:
This algorithm is exactly `adjacent_find` but for some reason `thrust` doesn't have this algorithm? :disappointed: I would recommend moving `prev_ch` to the capture list. I am also confused why the lambda doesn't have to be marked `mutable` - seeing as we are mutating `prev_ch`

Reply:
Not sure why I didn't realize this immediately but although `thrust` doesn't have `adjacent_find` - it does have the more general version of it `mismatch`. So this code can be replaced with:
```cpp
  auto const pair = thrust::mismatch(
    thrust::seq, key_begin, end - 1, key_begin + 1, [quotechar] __device__(auto prev_ch, auto ch) {
      return not(ch == quotechar && prev_ch != '\\');
    });

  return {key_begin, pair.first};
```
At which point I think you have drop the `thrust::seq` as well :slightly_smiling_face: